### PR TITLE
feat(aws_backend_role): Implement controller for AWS backend role

### PR DESCRIPTION
**Pull request by RonnanSouza**

### CODE REVIEW

File: apis/aws/v1alpha1/role_types.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	Backend               string   `json:"authBackend" binding:"required"`
+	CredentialType        string   `json:"credentialType" binding:"required"`
+	IamRolesArn           []string `json:"iamRoles" binding:"required"`
+	PoliciesArn           []string `json:"policiesArn" binding:"required"`
+	PolicyDocument        string   `json:"policyDocument" binding:"required"`
+	IamGroups             []string `json:"iamGroups" binding:"required"`
+	UserPath              string   `json:"userPath" binding:"required"`
+	PermissionBoundaryArn string   `json:"permissionsBoundaryArn" binding:"required"`
+	DefaultStsTTL         int      `json:"defaultStsTtl" binding:"required"`
+	MaxStsTTL             int      `json:"maxStsTtl" binding:"required"`
```
Comment (RonnanSouza) [inline apis/aws/v1alpha1/role_types.go]: Hey Brenno, I would add docs to all the fields here. This documentation will be put into the CRDs generated. Also, I think the `binding:"required"` is useless in this scenario. Just make sure you add the optional fields as pointers and put the `// +optional` comment on top of it. For those things, you may check how aws-provider does for example or even check my branch.

File: internal/controller/role/role.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+// Create an AWS Secret Backend Role
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	role, ok := mg.(*v1alpha1.Role)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotRole)
+	}
+
+	_, err := c.writeRole(role)
+	if err != nil {
+		return managed.ExternalCreation{}, fmt.Errorf("error creating role")
```
Comment (RonnanSouza) [inline internal/controller/role/role.go]: Roberto has implemented some default error types. I think we should leverage it and wrap the error with that one.

**Review by RonnanSouza — COMMENTED** [review]
Hey @beuleal , I've added some comments about things that I noticed and I think you're on the way.

File: internal/controller/role/role.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+	name := role.Name
+	authBackend := role.Spec.ForProvider.Backend
+
+	path := authBackend + "/roles/" + name
+
+	c.logger.Debug("[DEBUG] Reading role from %q", path)
+	_, err := c.client.Logical().Read(path)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("error reading role %q: %s", path, err)
```
Comment (Kasama) [inline internal/controller/role/role.go]: always wrap the returned errors, following [this guidelines](https://github.com/crossplane/crossplane/blob/master/CONTRIBUTING.md#wrap-errors) and use a `const` defined error message

Comment (Kasama) [reply internal/controller/role/role.go]: this applies to all other returned errors as well

File: internal/controller/role/role.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	}
+	log.Printf("[DEBUG] Read role from %q", path)
+
+	return managed.ExternalObservation{
```
Comment (Kasama) [inline internal/controller/role/role.go:168]: don't forget to [`setCondition(Available)`](https://github.com/topfreegames/crossplane-provider-vault/blob/7accb7f38b2e18aed61cf1a7b82349b03b5c9486/internal/controller/policy/policy.go#L136) to set the `Ready` condition when the observed result is both existent and up-to-date.

File: apis/aws/v1alpha1/role_types.go (hunk 1 above)
Comment (Kasama) [inline apis/aws/v1alpha1/role_types.go]: I don't think all of these parameters should be `required`. I think following terraform's defaults is sensible

Comment (beuleal) [reply apis/aws/v1alpha1/role_types.go]: It was a draft while I was developing. It's fixed already.

File: internal/controller/role/role.go (hunk 3 above)
Comment (beuleal) [reply internal/controller/role/role.go]: Fixed! Thanks!

Comment (beuleal) [reply internal/controller/role/role.go]: Thank you for sharing the guidelines! Really appreciate!

File: internal/controller/role/vault.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+// Note: these values comes from https://registry.terraform.io/providers/hashicorp/vault/latest/docs/resources/aws_secret_backend_role
+
+const (
+	errMarshal   = "error while parsing to json"
+	errUnmarshal = "error parsing json to interface"
+)
+
+// CrossplaneToVault is a transport object to send to vault. The reason we are using it, its because vault only accepts values as snake_case
+type CrossplaneToVault struct {
```
Comment (RonnanSouza) [inline internal/controller/role/vault.go]: I would use something like `Role` here.

File: internal/controller/role/role.go [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+	}
+	c.logger.Debug("Deleted role %q on AWS backend %q", name, authBackend)
+
+	return nil
+}
+
+// validate the role as some fields are only allowed in case another field has a different value
+//
+//nolint:gocyclo // Disabling lint here as we are doing many validation between fields that could not be done using kubebuilder
+func validate(role *v1alpha1.Role) error {
```
Comment (RonnanSouza) [inline internal/controller/role/role.go]: I would put this as a method in the `CrossplaneToVault` with several "private" methods checking each field. THis you'd reduce the complexity lint issue.

**Review by RonnanSouza — COMMENTED** [review]
Hey Brenno, I've added couple comments about things we discussed but the overall is a pretty good job. Congratz, mate!

File: internal/controller/role/role.go [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+		if defaultStsTTL > 0 {
+			return errors.New(errDefaultSts)
+		}
+	}
+
+	return nil
+}
+
+// isUpToDate checks if both data are the same to set Ready as true in Crossplane
+func isUpToDate(crossplaneData, vaultData CrossplaneToVault) (bool, error) {
```
Comment (RonnanSouza) [inline internal/controller/role/role.go]: This function should only return a boolean

**Review by Kasama — APPROVED** [review]

**Pull request merged**